## Classification using Deep Learning
### Amir, 2025, NUS SoC ACE
#### .
#### We use the heart disease dataset with 2 classes.


In [2]:
# first neural network with keras tutorial
from numpy import loadtxt
from keras.models import Sequential
from keras.layers import Dense , Dropout
import numpy as np

In [3]:
# mounting the Google drive
#from google.colab import drive

#drive.mount('/content/drive')

# Load the dataset
fp = "sample_data/datasets_heart_noheader.csv"

In [4]:
# load the dataset
dataset = loadtxt(fp, delimiter=',')
# split into input (X) and output (y) variables
print(dataset.shape)
X = dataset[0:270,0:13]
print(X.shape)
y = dataset[0:270,13]
X2 = dataset[270:302,0:13]
y2 = dataset[270:302,13]
print(X2)
print(y2)

(303, 14)
(270, 13)
[[5.80e+01 1.00e+00 0.00e+00 1.14e+02 3.18e+02 0.00e+00 2.00e+00 1.40e+02
  0.00e+00 4.40e+00 0.00e+00 3.00e+00 1.00e+00]
 [5.80e+01 0.00e+00 0.00e+00 1.70e+02 2.25e+02 1.00e+00 0.00e+00 1.46e+02
  1.00e+00 2.80e+00 1.00e+00 2.00e+00 1.00e+00]
 [6.70e+01 1.00e+00 2.00e+00 1.52e+02 2.12e+02 0.00e+00 0.00e+00 1.50e+02
  0.00e+00 8.00e-01 1.00e+00 0.00e+00 3.00e+00]
 [4.40e+01 1.00e+00 0.00e+00 1.20e+02 1.69e+02 0.00e+00 1.00e+00 1.44e+02
  1.00e+00 2.80e+00 0.00e+00 0.00e+00 1.00e+00]
 [6.30e+01 1.00e+00 0.00e+00 1.40e+02 1.87e+02 0.00e+00 0.00e+00 1.44e+02
  1.00e+00 4.00e+00 2.00e+00 2.00e+00 3.00e+00]
 [6.30e+01 0.00e+00 0.00e+00 1.24e+02 1.97e+02 0.00e+00 1.00e+00 1.36e+02
  1.00e+00 0.00e+00 1.00e+00 0.00e+00 2.00e+00]
 [5.90e+01 1.00e+00 0.00e+00 1.64e+02 1.76e+02 1.00e+00 0.00e+00 9.00e+01
  0.00e+00 1.00e+00 1.00e+00 2.00e+00 1.00e+00]
 [5.70e+01 0.00e+00 0.00e+00 1.40e+02 2.41e+02 0.00e+00 1.00e+00 1.23e+02
  1.00e+00 2.00e-01 1.00e+00 0.00e+00 3.00e+00]
 [4.

In [36]:
# define the keras model
from tensorflow.keras import regularizers
model = Sequential()
model.add(Dense(12, input_dim=13, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(4, activation='relu', kernel_regularizer=regularizers.l2(0.01)))
model.add(Dropout(0.5))
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [37]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [38]:
# fit the keras model on the dataset
model.fit(X, y, epochs=600, batch_size=10)

Epoch 1/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.5478 - loss: 2.0604
Epoch 2/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4891 - loss: 0.9918
Epoch 3/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4785 - loss: 0.7729
Epoch 4/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5068 - loss: 0.7405
Epoch 5/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6147 - loss: 0.6997
Epoch 6/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5812 - loss: 0.7207
Epoch 7/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5659 - loss: 0.6862
Epoch 8/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5555 - loss: 0.7299
Epoch 9/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6215 - loss: 0.6619
Epoch 10/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6335 - loss: 0.6995
Epoch 11/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5729 - loss: 0.7096
Epoch 12/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy:

In [39]:
# evaluate the keras model
_, accuracy = model.evaluate(X, y)
print('Accuracy: %.2f' % (accuracy*100))

9/9 ━━━━━━━━━━━━━━━━━━━━ 1s 52ms/step - accuracy: 0.7918 - loss: 0.4788
Accuracy: 82.22


In [40]:
# make probability predictions with the model
predictions = model.predict(X)
# round predictions
rounded = [round(x[0]) for x in predictions]

9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step


In [41]:
# make class predictions with the model
for i in range(15):
	print('%s => %d (expected %d)' % (X[i].tolist(), rounded[i], y[i]))

[63.0, 1.0, 3.0, 145.0, 233.0, 1.0, 0.0, 150.0, 0.0, 2.3, 0.0, 0.0, 1.0] => 1 (expected 1)
[37.0, 1.0, 2.0, 130.0, 250.0, 0.0, 1.0, 187.0, 0.0, 3.5, 0.0, 0.0, 2.0] => 0 (expected 1)
[41.0, 0.0, 1.0, 130.0, 204.0, 0.0, 0.0, 172.0, 0.0, 1.4, 2.0, 0.0, 2.0] => 1 (expected 1)
[56.0, 1.0, 1.0, 120.0, 236.0, 0.0, 1.0, 178.0, 0.0, 0.8, 2.0, 0.0, 2.0] => 1 (expected 1)
[57.0, 0.0, 0.0, 120.0, 354.0, 0.0, 1.0, 163.0, 1.0, 0.6, 2.0, 0.0, 2.0] => 1 (expected 1)
[57.0, 1.0, 0.0, 140.0, 192.0, 0.0, 1.0, 148.0, 0.0, 0.4, 1.0, 0.0, 1.0] => 1 (expected 1)
[56.0, 0.0, 1.0, 140.0, 294.0, 0.0, 0.0, 153.0, 0.0, 1.3, 1.0, 0.0, 2.0] => 1 (expected 1)
[44.0, 1.0, 1.0, 120.0, 263.0, 0.0, 1.0, 173.0, 0.0, 0.0, 2.0, 0.0, 3.0] => 1 (expected 1)
[52.0, 1.0, 2.0, 172.0, 199.0, 1.0, 1.0, 162.0, 0.0, 0.5, 2.0, 0.0, 3.0] => 1 (expected 1)
[57.0, 1.0, 2.0, 150.0, 168.0, 0.0, 1.0, 174.0, 0.0, 1.6, 2.0, 0.0, 2.0] => 1 (expected 1)
[54.0, 1.0, 0.0, 140.0, 239.0, 0.0, 1.0, 160.0, 0.0, 1.2, 2.0, 0.0, 2.0] => 0 (expected 1)

In [42]:
M= len(rounded)
print( 'Accuracy= ' , 1- np.sum(np.abs( rounded - y )) / M)

Accuracy=  0.8222222222222222


In [43]:
# evaluate the keras model, testing
_, accuracy = model.evaluate(X2, y2)
print('Accuracy: %.2f' % (accuracy*100))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 105ms/step - accuracy: 0.7500 - loss: 0.5142
Accuracy: 75.00


In [44]:
# make probability predictions with the model, testing
predictions = model.predict(X2)
# round predictions
rounded = [round(x[0]) for x in predictions]

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step


In [45]:
# make class predictions with the model, testing
for i in range(25):
	print('%s => %d (expected %d)' % (X2[i].tolist(), rounded[i], y2[i]))

[58.0, 1.0, 0.0, 114.0, 318.0, 0.0, 2.0, 140.0, 0.0, 4.4, 0.0, 3.0, 1.0] => 0 (expected 0)
[58.0, 0.0, 0.0, 170.0, 225.0, 1.0, 0.0, 146.0, 1.0, 2.8, 1.0, 2.0, 1.0] => 0 (expected 0)
[67.0, 1.0, 2.0, 152.0, 212.0, 0.0, 0.0, 150.0, 0.0, 0.8, 1.0, 0.0, 3.0] => 1 (expected 0)
[44.0, 1.0, 0.0, 120.0, 169.0, 0.0, 1.0, 144.0, 1.0, 2.8, 0.0, 0.0, 1.0] => 0 (expected 0)
[63.0, 1.0, 0.0, 140.0, 187.0, 0.0, 0.0, 144.0, 1.0, 4.0, 2.0, 2.0, 3.0] => 0 (expected 0)
[63.0, 0.0, 0.0, 124.0, 197.0, 0.0, 1.0, 136.0, 1.0, 0.0, 1.0, 0.0, 2.0] => 1 (expected 0)
[59.0, 1.0, 0.0, 164.0, 176.0, 1.0, 0.0, 90.0, 0.0, 1.0, 1.0, 2.0, 1.0] => 0 (expected 0)
[57.0, 0.0, 0.0, 140.0, 241.0, 0.0, 1.0, 123.0, 1.0, 0.2, 1.0, 0.0, 3.0] => 0 (expected 0)
[45.0, 1.0, 3.0, 110.0, 264.0, 0.0, 1.0, 132.0, 0.0, 1.2, 1.0, 0.0, 3.0] => 1 (expected 0)
[68.0, 1.0, 0.0, 144.0, 193.0, 1.0, 1.0, 141.0, 0.0, 3.4, 1.0, 2.0, 3.0] => 0 (expected 0)
[57.0, 1.0, 0.0, 130.0, 131.0, 0.0, 1.0, 115.0, 1.0, 1.2, 1.0, 1.0, 3.0] => 0 (expected 0)


In [26]:
M= len(rounded)
print( 'Accuracy= ' , 1- np.sum(np.abs(rounded - y2)) / M)

Accuracy=  0.75


## Practices for You
1. Run the program a few times. Is there any meaningful difference between the results? Why?
  - Yes, even if the neural network (NN) architecture is not changed and the dataset is exactly the same, running the training process again can and often will change the model performance.
While the evaluation of a pre-trained, frozen model on the same data should produce identical results, the training process itself is stochastic (non-deterministic).
Here is a breakdown of why this happens and how to manage it:
Why Results Change (Sources of Randomness)
Random Weight Initialization: Neural networks start by initializing their internal weights with random numbers. Different initializations lead to different paths during optimization, resulting in different final models.
Data Shuffling: Data is usually shuffled at the start of each epoch to improve learning. If not strictly controlled, this means the model sees examples in a different order each time, affecting weight updates.
GPU Non-determinism: Many GPU operations (e.g., in CUDA/PyTorch/TensorFlow) are parallelized and not strictly deterministic. The order of operations may vary slightly, leading to different floating-point results that accumulate over training.
Hardware/Numerical Differences: Small differences in how CPUs or GPUs handle floating-point math can result in different, yet equally valid, performance metrics.
2. Train the model for 1100 epochs. what are the training and testing accuracy?
  - Training - 86.3, Testing - 75
3. repeat that with 6000 training epochs. what are the training and testing accuracy?
- Training - 90, Testing - 75
4. move back to 600 training epochs. add two dense layers with optional features. what are the training and testing accuracy?
  - adding more layers and optional features such as dropout amd regularizer in this case did not improve accuracy. Training accuracy is 82%
5. go for 6000 training epochs. Do you feel any overfitting?
6. set back everything to the original setting, then try to normalize the training and testing features and outputs using standard_scaler. Then see if anything gets better or not.

#### points:
- Use GPU as your runtime type